diff --git a/build/core/config.mk b/build/core/config.mk
index 83af311..00cc838 100644
--- a/build/core/config.mk
+++ b/build/core/config.mk
@@ -351,10 +351,10 @@ ifeq ($(HOST_OS),darwin)
 HOST_JDK_TOOLS_JAR :=
 else
 HOST_JDK_TOOLS_JAR:= $(shell $(BUILD_SYSTEM)/find-jdk-tools-jar.sh)
-ifeq ($(wildcard $(HOST_JDK_TOOLS_JAR)),)
-$(error Error: could not find jdk tools.jar, please install JDK6, \
-    which you can download from java.sun.com)
-endif
+#ifeq ($(wildcard $(HOST_JDK_TOOLS_JAR)),)
+#$(error Error: could not find jdk tools.jar, please install JDK6, \
+#    which you can download from java.sun.com)
+#endif
 endif
 
 # Is the host JDK 64-bit version?
