--- a/kernel/softwinner/907/drivers/input/touchscreen/Kconfig	2013-04-13 21:39:48.856570204 +0200
+++ b/kernel/softwinner/907/drivers/input/touchscreen/Kconfig	2013-04-13 21:51:24.546988879 +0200
@@ -738,6 +738,13 @@
        help
           GT801 touchscreen driver
 
+config TOUCHSCREEN_GT801NI
+       tristate "hack to load binary module for GT801NI"
+       depends on I2C
+       help
+         This implements a symbol to be able to load a
+         binary module for GT801NI touchscreen
+
 config  TOUCHSCREEN_GT811
        tristate  "dpt microelctronics gt811 touchscreen driver"
        depends on I2C
--- a/kernel/softwinner/907/drivers/input/touchscreen/Makefile	2013-04-13 21:39:48.857570331 +0200
+++ b/kernel/softwinner/907/drivers/input/touchscreen/Makefile	2013-04-13 22:12:14.081798328 +0200
@@ -64,6 +64,7 @@
 obj-$(CONFIG_TOUCHSCREEN_FT5X_TS) += ft5x_ts.o
 obj-$(CONFIG_TOUCHSCREEN_SSD253X_TS) 	+= ssd253x-ts.o
 obj-$(CONFIG_TOUCHSCREEN_GT801)         += goodix_touch.o
+obj-$(CONFIG_TOUCHSCREEN_GT801NI)	+= goodix_hack.o
 obj-$(CONFIG_TOUCHSCREEN_GT811)         += gt811_ts.o
 obj-$(CONFIG_TOUCHSCREEN_GT818)         += gt818_ts.o
 obj-$(CONFIG_TOUCHSCREEN_ZT8031)         += zt8031.o
--- /dev/null
+++ b/kernel/softwinner/907/drivers/input/touchscreen/goodix_hack.c
@@ -0,0 +1,5 @@
+#include <linux/module.h>
+
+int m_inet_ctpState = 0;
+
+EXPORT_SYMBOL(m_inet_ctpState);
-- 
1.7.10
