commit c8a3a51e77ea1fd82ebe47500904be7260de4d33
From: ugers <yugers@gmail.com>
Date:   Fri Mar 27 23:29:03 2015 +0300 
Subject: [PATCH] vendor_cm

Change-Id: Id14753791b3038ee73835e0898f82e2c1ff7bb64

diff --git a/config/common.mk b/config/common.mk
index 8fce1d2..e965866 100644
--- a/config/common.mk
+++ b/config/common.mk
@@ -75,7 +75,7 @@ PRODUCT_PROPERTY_OVERRIDES += persist.sys.dun.override=0
 
 ifneq ($(TARGET_BUILD_VARIANT),eng)
 # Enable ADB authentication
-ADDITIONAL_DEFAULT_PROPERTIES += ro.adb.secure=1
+#ADDITIONAL_DEFAULT_PROPERTIES += ro.adb.secure=1
 endif
 
 # Copy over the changelog to the device
